text/0033-input-versions.md

```diff
@@ -0,0 +1,209 @@
+
+## Drawbacks
+
+Implementing this RFC will add some complexity to the various Cedar parsers. They will now have to accept inputs that have a version annotation, and those that do not. They will have to minimally use that annotation to improve functionality and/or error messages. Doing this requires thinking carefully about the meaning of an annotation (e.g., what span of versions use the same format).
+
+The presence of a version annotation may imply more than this RFC is actually offering, leading to customer confusion or disappointment. For example, customers may expect that the current parser can parse and properly interpret all formats up to the present one, rather than reject those earlier formats (gracefully).
+
+Adding an annotation adds some maintenance burden on customers. When a version annotation is present in a policy file, if the policies are changed to use a newer version of Cedar, the version annotation needs to be changed; it is not hard to forget to do so.
+
+If an input format changes in a way that older syntax is interpreted differently, such as the `appliesTo` in the JSON schema as per the request in [issue 351](https://github.com/cedar-policy/cedar/issues/351), the fact that a version annotation is optional presents a problem: A newer and older parser could both accept the same input, but interpret it differently. If a version annotation was required, such an ambiguous interpretation would not be possible.
+
+## Alternatives
+
+### Do nothing to Cedar; version info can appear in application metadata
+
+Rather than add a version annotation in the Cedar input syntax directly, version information can just be stored in metadata. For example, an application could store Cedar schemas in a document database and use a format like
+```
+{
+    "cedar_verson": "3.1.0",
+    "schema" : {
+        "foo::bar" {
+            "entityTypes" : { ... },
+            "actions": { ... }
+        }
+    }
+}
+```
+Such annotations could then trigger different application actions, e.g., calling different versions of the Cedar evaluator.
+
+Versioning is especially useful when the same syntax is interpreted differently in later versions. This situation should be (very) rare, though. When it happens, the newer parser can warn about the change in syntax, especially in the case of errors that would trigger due to the change.
+
+### Change, or drop, the policy annotation format
+
+The use of `;` to distinguish per-policy annotations from file-wide policy annotations may be too subtle. Users could mistakenly write the following:
+```
+@cedar_version("2.4.0")
+permit(
+    principal == User::"Alice",
+    action == Action::"viewPhoto",
+    resource == Photo::"vacation97.jpg"
+);
+```
+They might think that the version applies to the whole file, but the lack of a semi-colon means the annotation just applies to the given policy.
```

Comment:
I didn't notice the semicolon when this syntax was first presented, and assumed the version applied on a per-policy basis. So I was confused when the RFC said later that it applied to the whole file. If we decide to keep policy annotations, then I'd vote for a different syntax like `#cedar_version(...)`.

I like the idea of only using versions for entity data & schemas. But, like this RFC says, that implies that the policy syntax is always backwards compatible. Unfortunately, that's not the case due to [RFC20](https://github.com/cedar-policy/rfcs/blob/input-versions/text/0020-unique-record-keys.md).

Reply:
I've been trying to figure out why am I in favor of including version into entity data and schemas but opposed to doing the same for policies. I have a hunch:

* In most products built on Cedar schema and entity data would be represented as single 'document'. Entire schema is a single JSON file and same goes for request data.
* On the other hand, policies are going to be split across multiple documents in which case we either need to annotate each policy (not practical) or have version of the policy set stored in the config somewhere. 

Each product tailors policy management to their own unique need and having inline annotation of the version for policies seems to be less ideal in that case.

Reply:
If users manage their own policy versioning, why would they not want to manage their schema and entity versioning too?

I like @khieta's solution from [this comment](https://github.com/cedar-policy/rfcs/pull/33#discussion_r1381889216), which is to put all versioning information into a single configuration file, rather than spread it across many files (schemas, entity data, policies, etc).